app/src/main/java/com/github/studydistractor/sdp/account/FirebaseUser.kt

```diff
@@ -0,0 +1,51 @@
+package com.github.studydistractor.sdp.account
+
+import android.util.Log
+import androidx.compose.runtime.mutableStateListOf
+import androidx.compose.runtime.snapshots.SnapshotStateList
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.database.DataSnapshot
+import com.google.firebase.database.DatabaseError
+import com.google.firebase.database.FirebaseDatabase
+import com.google.firebase.database.ValueEventListener
+
+class FirebaseUser : UserInterface {
+    private val auth = FirebaseAuth.getInstance()
+    private val db = FirebaseDatabase.getInstance()
+
+    private val FRIENDSPATH = "Friends"
+
+
+    override fun getCurrentUid(): String? {
+        return auth.uid
+    }
+
+    override fun addNewFriend(uid1: String, uid2: String) {
+        db.getReference(FRIENDSPATH).child(uid1).child(uid2).setValue(uid2)
+        db.getReference(FRIENDSPATH).child(uid2).child(uid1).setValue(uid2)
+    }
+
+    override fun removeFriend(uid1: String, uid2: String) {
+        db.getReference(FRIENDSPATH).child(uid1).child(uid2).removeValue()
+        db.getReference(FRIENDSPATH).child(uid2).child(uid1).removeValue()
+    }
+
+    override fun fetchAllFriends(uid: String): SnapshotStateList<String> {
+        val databaseRef = db.getReference(FRIENDSPATH).child(uid)
+        val result =  mutableStateListOf<String>()
+        databaseRef.addListenerForSingleValueEvent(object : ValueEventListener {
```

Comment:
I know I made the same mistake but now we need to add the listener in init { } and put the `val result` as attribute of the class. and the fetch function only returns the reference to result. This way the result observe the changes only once and the result is updated only when Firebase tells it to.